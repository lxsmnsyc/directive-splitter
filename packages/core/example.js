import { compile } from 'dismantle';

const code = `
import { server$ } from 'my-example';

const foo = 'foo';
const bar = 'bar';

function log(prefix, suffix) {
  console.log(prefix, suffix);
}

const logPrefix = server$((prefix) => {
  if (prefix === 'foo') {
    log(prefix, foo);
  } else {
    log(prefix, bar);
  }
});
`;

const result = await compile(code, '/path/to/example.ts', {
  key: 'example',
  mode: 'server',
  env: 'development',
  directives: [
    {
      value: 'use server',
      import: {
        kind: 'named',
        name: '$$server',
        source: 'my-example',
      },
    },
  ],
  functions: [
    {
      source: {
        kind: 'named',
        name: 'server$',
        source: 'my-example',
      },
      target: {
        kind: 'named',
        name: 'server$',
        source: 'my-example/server',
      },
    },
  ],
});

console.log('ENTRY');
console.log(result.code);
console.log(result.map);

console.log('\nVIRTUALS');
for (const [name, content] of result.files) {
  console.log('FILE', name);
  console.log(content);
}

console.log('\nENTRIES');
console.log(result.entries);

console.log('\nROOTs');
console.log(result.roots);
