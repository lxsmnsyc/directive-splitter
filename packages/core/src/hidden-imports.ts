import type { ImportDefinition } from './types';

export const HIDDEN_FUNC: ImportDefinition = {
  kind: 'named',
  source: 'dismantle/runtime',
  name: '$$func',
};

export const HIDDEN_GENERATOR: ImportDefinition = {
  kind: 'named',
  source: 'dismantle/runtime',
  name: '$$gen',
};
