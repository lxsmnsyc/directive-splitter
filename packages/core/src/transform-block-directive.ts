import type * as babel from '@babel/core';
import type * as t from '@babel/types';
import { splitBlockDirective } from './split-block-directive';
import type { BlockDirectiveDefinition, StateContext } from './types';
import {
  cleanBlockForDirectives,
  cleanBlockForFauxDirectives,
  getDefinitionFromDirectives,
  getDefinitionFromFauxDirectives,
} from './utils/directive-check';

function getBlockDirectiveDefinition(
  ctx: StateContext,
  path: babel.NodePath<t.BlockStatement>,
): BlockDirectiveDefinition | undefined {
  const definition = getDefinitionFromDirectives(ctx, 'block-directive', path);
  if (definition) {
    cleanBlockForDirectives(path, definition);
    return definition;
  }
  const fauxDefinition = getDefinitionFromFauxDirectives(
    ctx,
    'block-directive',
    path,
  );
  if (fauxDefinition) {
    cleanBlockForFauxDirectives(path, fauxDefinition);
    return fauxDefinition;
  }
  return undefined;
}

export function transformBlockDirective(
  ctx: StateContext,
  path: babel.NodePath<t.BlockStatement>,
): void {
  const definition = getBlockDirectiveDefinition(ctx, path);
  if (definition) {
    splitBlockDirective(ctx, path, definition);
  }
}
