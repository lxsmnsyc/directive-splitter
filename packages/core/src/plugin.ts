import type * as babel from '@babel/core';
import { transformCall } from './transform-call';
import { transformBlockDirective } from './transform-block-directive';
import type { StateContext } from './types';
import { registerImportSpecifiers } from './utils/register-import-specifiers';

function bubbleFunctionDeclaration(): void {

}

interface State extends babel.PluginPass {
  opts: StateContext;
}

export function plugin(): babel.PluginObj<State> {
  return {
    name: 'dismantle',
    visitor: {
      Program(path, ctx) {
        registerImportSpecifiers(ctx.opts, path);
      },
      ArrowFunctionExpression(path, ctx) {

      },
      FunctionExpression(path, ctx) {

      },
      FunctionDeclaration(path, ctx) {

      },
      BlockStatement(path, ctx) {
        transformBlockDirective(ctx.opts, path);
      },
      CallExpression(path, ctx) {
        transformCall(ctx.opts, path);
      },
    },
  };
}
