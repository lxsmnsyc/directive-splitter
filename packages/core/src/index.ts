import * as babel from '@babel/core';
import path from 'path';
import type { Options, StateContext } from './types';
import assert from './utils/assert';
import { plugin } from './plugin';
import xxHash32 from './utils/xxhash32';

export interface Output {
  code: babel.BabelFileResult['code'];
  map: babel.BabelFileResult['map'];
  files: Map<string, string>;
}

export {
  NamedImportDefinition,
  DefaultImportDefinition,
  DirectiveDefinition,
  ImportDefinition,
  Options,
} from './types';

export async function compile(
  code: string,
  id: string,
  options: Options,
): Promise<Output> {
  const parsedPath = path.parse(id);

  const ctx: StateContext = {
    path: parsedPath,
    imports: new Map(),
    virtual: {
      files: new Map(),
      count: 0,
    },
    options,
    bindings: new Map(),
    blocks: {
      hash: xxHash32(id).toString(16),
      count: 0,
    },
  };

  const plugins: babel.ParserOptions['plugins'] = ['jsx'];

  if (/\.[mc]?tsx?$/i.test(id)) {
    plugins.push('typescript');
  }

  const result = await babel.transformAsync(code, {
    plugins: [[plugin, ctx]],
    parserOpts: {
      plugins,
    },
    filename: parsedPath.base,
    ast: false,
    sourceFileName: id,
    sourceMaps: true,
    configFile: false,
    babelrc: false,
  });

  assert(result, 'invariant');

  const files = new Map();

  for (const [key, value] of ctx.virtual.files) {
    files.set(path.join(parsedPath.dir, key), value);
  }

  return {
    code: result.code,
    map: result.map,
    files,
  };
}
