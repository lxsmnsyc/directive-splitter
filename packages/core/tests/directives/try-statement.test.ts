import { describe, it, expect } from 'vitest';
import * as compiler from '../../src';
import { ID, CLIENT, SERVER } from './example';

describe('TryStatement', () => {
  describe('client', () => {
    it('should transform valid server try statements', async () => {
      const code = `
      try {
        'use server';
        await doStuff();
      } catch (err) {
        report(err);
      }
      `;
      expect(await compiler.compile(ID, code, CLIENT)).toMatchSnapshot();
    });
    it('should skip server try statements in non-async functions', async () => {
      const code = `
      const example = () => {
        try {
          'use server';
          doStuff();
        } catch (err) {
          report(err);
        }
      };
      `;
      expect(await compiler.compile(ID, code, CLIENT)).toMatchSnapshot();
    });
    it('should transform valid server functions with scope', async () => {
      const code = `
      async function foo() {
        const value = 'foo bar';
        try {
          'use server';
          await doStuff(value);
        } catch (err) {
          report(err);
        }
      }
      `;
      expect(await compiler.compile(ID, code, CLIENT)).toMatchSnapshot();
    });
    it('should skip top-level values for scope', async () => {
      const code = `
      const value = 'foo bar';
      try {
        'use server';
        await doStuff(value);
      } catch (err) {
        report(err);
      }
      `;
      expect(await compiler.compile(ID, code, CLIENT)).toMatchSnapshot();
    });
  });
  describe('server', () => {
    it('should transform valid server try statements', async () => {
      const code = `
      try {
        'use server';
        await doStuff();
      } catch (err) {
        report(err);
      }
      `;
      expect(await compiler.compile(ID, code, SERVER)).toMatchSnapshot();
    });
    it('should skip server try statements in non-async functions', async () => {
      const code = `
      const example = () => {
        try {
          'use server';
          doStuff();
        } catch (err) {
          report(err);
        }
      };
      `;
      expect(await compiler.compile(ID, code, SERVER)).toMatchSnapshot();
    });
    it('should transform valid server functions with scope', async () => {
      const code = `
      async function foo() {
        const value = 'foo bar';
        try {
          'use server';
          await doStuff(value);
        } catch (err) {
          report(err);
        }
      }
      `;
      expect(await compiler.compile(ID, code, SERVER)).toMatchSnapshot();
    });
    it('should skip top-level values for scope', async () => {
      const code = `
      const value = 'foo bar';
      try {
        'use server';
        await doStuff(value);
      } catch (err) {
        report(err);
      }
      `;
      expect(await compiler.compile(ID, code, SERVER)).toMatchSnapshot();
    });
  });
});
