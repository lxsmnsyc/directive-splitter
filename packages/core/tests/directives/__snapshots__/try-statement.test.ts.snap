// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TryStatement > client > should skip server try statements in non-async functions 1`] = `
{
  "code": "const example = () => {
  try {
    'use server';
    doStuff();
  } catch (err) {
    report(err);
  }
};",
  "entries": [],
  "files": Map {},
  "map": {
    "file": undefined,
    "mappings": "AACM,MAAMA,OAAO,GAAGA,CAAA,KAAM;EACpB,IAAI;IACF,YAAY;IACZC,OAAO,CAAC,CAAC;EACX,CAAC,CAAC,OAAOC,GAAG,EAAE;IACZC,MAAM,CAACD,GAAG,CAAC;EACb;AACF,CAAC",
    "names": [
      "example",
      "doStuff",
      "err",
      "report",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      const example = () => {
        try {
          'use server';
          doStuff();
        } catch (err) {
          report(err);
        }
      };
      ",
    ],
    "version": 3,
  },
  "roots": [],
}
`;

exports[`TryStatement > client > should skip top-level values for scope 1`] = `
{
  "code": "import { $$func as _$$func } from "dismantle/runtime";
import { value } from "./example.ts?my-example=0.ts";
try {
  const [_type, _result] = await _$$func((await import("./example.ts?my-example=2.ts")).default, null)();
} catch (err) {
  report(err);
}",
  "entries": [
    "/path/to/example.ts?my-example=2.ts",
  ],
  "files": Map {
    "/path/to/example.ts?my-example=0.ts" => {
      "code": "export const value = 'foo bar';",
      "map": {
        "file": undefined,
        "mappings": "aACYA,KAAK,GAAG,SAAS",
        "names": [
          "value",
        ],
        "sourceRoot": undefined,
        "sources": [
          "/path/to/example.ts",
        ],
        "sourcesContent": [
          null,
        ],
        "version": 3,
      },
    },
    "/path/to/example.ts?my-example=1.ts" => {
      "code": "export default (async function () {
  try {
    await doStuff(value);
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
      "map": {
        "file": undefined,
        "mappings": ";;IAIQ,MAAMA,OAAO,CAACC,KAAK,CAAC;EAAC,SAAAC,MAAA;IAAA,WAAAA,MAAA;EAAA;EAAA;AAAA",
        "names": [
          "doStuff",
          "value",
          "_error",
        ],
        "sourceRoot": undefined,
        "sources": [
          "/path/to/example.ts",
        ],
        "sourcesContent": [
          null,
        ],
        "version": 3,
      },
    },
    "/path/to/example.ts?my-example=2.ts" => {
      "code": "import { server as _entry } from "my-example/client";
export default _entry("26ecc3a5-0-anonymous");",
      "map": {
        "file": undefined,
        "mappings": "",
        "names": [],
        "sourceRoot": undefined,
        "sources": [],
        "sourcesContent": [],
        "version": 3,
      },
    },
  },
  "map": {
    "file": undefined,
    "mappings": ";;AAEM;EAAA,OAAAA,KAAA,EAAAC,OAAA,UAAAC,OAAA,gDAAAC,OAAA;AAAA,EAGE,OAAOC,GAAG,EAAE;EACZC,MAAM,CAACD,GAAG,CAAC;AACb",
    "names": [
      "_type",
      "_result",
      "_$$func",
      "default",
      "err",
      "report",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      const value = 'foo bar';
      try {
        'use server';
        await doStuff(value);
      } catch (err) {
        report(err);
      }
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?my-example=1.ts",
  ],
}
`;

exports[`TryStatement > client > should transform valid server functions with scope 1`] = `
{
  "code": "import { $$func as _$$func } from "dismantle/runtime";
async function foo() {
  const value = 'foo bar';
  try {
    const [_type, _result] = await _$$func((await import("./example.ts?my-example=1.ts")).default, null)(value);
  } catch (err) {
    report(err);
  }
}",
  "entries": [
    "/path/to/example.ts?my-example=1.ts",
  ],
  "files": Map {
    "/path/to/example.ts?my-example=0.ts" => {
      "code": "export default (async function (value) {
  try {
    await doStuff(value);
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
      "map": {
        "file": undefined,
        "mappings": "gCAEcA,KAAK;EAAA;IAGT,MAAMC,OAAO,CAACD,KAAK,CAAC;EAAC,SAAAE,MAAA;IAAA,WAAAA,MAAA;EAAA;EAAA;AAAA",
        "names": [
          "value",
          "doStuff",
          "_error",
        ],
        "sourceRoot": undefined,
        "sources": [
          "/path/to/example.ts",
        ],
        "sourcesContent": [
          null,
        ],
        "version": 3,
      },
    },
    "/path/to/example.ts?my-example=1.ts" => {
      "code": "import { server as _entry } from "my-example/client";
export default _entry("26ecc3a5-0-foo");",
      "map": {
        "file": undefined,
        "mappings": "",
        "names": [],
        "sourceRoot": undefined,
        "sources": [],
        "sourcesContent": [],
        "version": 3,
      },
    },
  },
  "map": {
    "file": undefined,
    "mappings": ";AACM,eAAeA,GAAGA,CAAA,EAAG;EACnB,MAAMC,KAAK,GAAG,SAAS;EACvB;IAAA,OAAAC,KAAA,EAAAC,OAAA,UAAAC,OAAA,gDAAAC,OAAA,QADMJ,KAAK;EAAA,EAIT,OAAOK,GAAG,EAAE;IACZC,MAAM,CAACD,GAAG,CAAC;EACb;AACF",
    "names": [
      "foo",
      "value",
      "_type",
      "_result",
      "_$$func",
      "default",
      "err",
      "report",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      async function foo() {
        const value = 'foo bar';
        try {
          'use server';
          await doStuff(value);
        } catch (err) {
          report(err);
        }
      }
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?my-example=0.ts",
  ],
}
`;

exports[`TryStatement > client > should transform valid server try statements 1`] = `
{
  "code": "import { $$func as _$$func } from "dismantle/runtime";
try {
  const [_type, _result] = await _$$func((await import("./example.ts?my-example=1.ts")).default, null)();
} catch (err) {
  report(err);
}",
  "entries": [
    "/path/to/example.ts?my-example=1.ts",
  ],
  "files": Map {
    "/path/to/example.ts?my-example=0.ts" => {
      "code": "export default (async function () {
  try {
    await doStuff();
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
      "map": {
        "file": undefined,
        "mappings": ";;IAGQ,MAAMA,OAAO,CAAC,CAAC;EAAC,SAAAC,MAAA;IAAA,WAAAA,MAAA;EAAA;EAAA;AAAA",
        "names": [
          "doStuff",
          "_error",
        ],
        "sourceRoot": undefined,
        "sources": [
          "/path/to/example.ts",
        ],
        "sourcesContent": [
          null,
        ],
        "version": 3,
      },
    },
    "/path/to/example.ts?my-example=1.ts" => {
      "code": "import { server as _entry } from "my-example/client";
export default _entry("26ecc3a5-0-anonymous");",
      "map": {
        "file": undefined,
        "mappings": "",
        "names": [],
        "sourceRoot": undefined,
        "sources": [],
        "sourcesContent": [],
        "version": 3,
      },
    },
  },
  "map": {
    "file": undefined,
    "mappings": ";AACM;EAAA,OAAAA,KAAA,EAAAC,OAAA,UAAAC,OAAA,gDAAAC,OAAA;AAAA,EAGE,OAAOC,GAAG,EAAE;EACZC,MAAM,CAACD,GAAG,CAAC;AACb",
    "names": [
      "_type",
      "_result",
      "_$$func",
      "default",
      "err",
      "report",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      try {
        'use server';
        await doStuff();
      } catch (err) {
        report(err);
      }
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?my-example=0.ts",
  ],
}
`;

exports[`TryStatement > server > should skip server try statements in non-async functions 1`] = `
{
  "code": "const example = () => {
  try {
    'use server';
    doStuff();
  } catch (err) {
    report(err);
  }
};",
  "entries": [],
  "files": Map {},
  "map": {
    "file": undefined,
    "mappings": "AACM,MAAMA,OAAO,GAAGA,CAAA,KAAM;EACpB,IAAI;IACF,YAAY;IACZC,OAAO,CAAC,CAAC;EACX,CAAC,CAAC,OAAOC,GAAG,EAAE;IACZC,MAAM,CAACD,GAAG,CAAC;EACb;AACF,CAAC",
    "names": [
      "example",
      "doStuff",
      "err",
      "report",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      const example = () => {
        try {
          'use server';
          doStuff();
        } catch (err) {
          report(err);
        }
      };
      ",
    ],
    "version": 3,
  },
  "roots": [],
}
`;

exports[`TryStatement > server > should skip top-level values for scope 1`] = `
{
  "code": "import { $$func as _$$func } from "dismantle/runtime";
import { value } from "./example.ts?my-example=0.ts";
try {
  const [_type, _result] = await _$$func((await import("./example.ts?my-example=2.ts")).default, null)();
} catch (err) {
  report(err);
}",
  "entries": [
    "/path/to/example.ts?my-example=2.ts",
  ],
  "files": Map {
    "/path/to/example.ts?my-example=0.ts" => {
      "code": "export const value = 'foo bar';",
      "map": {
        "file": undefined,
        "mappings": "aACYA,KAAK,GAAG,SAAS",
        "names": [
          "value",
        ],
        "sourceRoot": undefined,
        "sources": [
          "/path/to/example.ts",
        ],
        "sourcesContent": [
          null,
        ],
        "version": 3,
      },
    },
    "/path/to/example.ts?my-example=1.ts" => {
      "code": "import { value } from "./example.ts?my-example=0.ts";
export default (async function () {
  try {
    await doStuff(value);
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
      "map": {
        "file": undefined,
        "mappings": ";;;IAIQ,MAAMA,OAAO,CAACC,KAAK,CAAC;EAAC,SAAAC,MAAA;IAAA,WAAAA,MAAA;EAAA;EAAA;AAAA",
        "names": [
          "doStuff",
          "value",
          "_error",
        ],
        "sourceRoot": undefined,
        "sources": [
          "/path/to/example.ts",
        ],
        "sourcesContent": [
          null,
        ],
        "version": 3,
      },
    },
    "/path/to/example.ts?my-example=2.ts" => {
      "code": "import { server as _entry } from "my-example/server";
import _root from "./example.ts?my-example=1.ts";
export default _entry("26ecc3a5-0-anonymous", _root);",
      "map": {
        "file": undefined,
        "mappings": "",
        "names": [],
        "sourceRoot": undefined,
        "sources": [],
        "sourcesContent": [],
        "version": 3,
      },
    },
  },
  "map": {
    "file": undefined,
    "mappings": ";;AAEM;EAAA,OAAAA,KAAA,EAAAC,OAAA,UAAAC,OAAA,gDAAAC,OAAA;AAAA,EAGE,OAAOC,GAAG,EAAE;EACZC,MAAM,CAACD,GAAG,CAAC;AACb",
    "names": [
      "_type",
      "_result",
      "_$$func",
      "default",
      "err",
      "report",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      const value = 'foo bar';
      try {
        'use server';
        await doStuff(value);
      } catch (err) {
        report(err);
      }
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?my-example=1.ts",
  ],
}
`;

exports[`TryStatement > server > should transform valid server functions with scope 1`] = `
{
  "code": "import { $$func as _$$func } from "dismantle/runtime";
async function foo() {
  const value = 'foo bar';
  try {
    const [_type, _result] = await _$$func((await import("./example.ts?my-example=1.ts")).default, null)(value);
  } catch (err) {
    report(err);
  }
}",
  "entries": [
    "/path/to/example.ts?my-example=1.ts",
  ],
  "files": Map {
    "/path/to/example.ts?my-example=0.ts" => {
      "code": "export default (async function (value) {
  try {
    await doStuff(value);
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
      "map": {
        "file": undefined,
        "mappings": "gCAEcA,KAAK;EAAA;IAGT,MAAMC,OAAO,CAACD,KAAK,CAAC;EAAC,SAAAE,MAAA;IAAA,WAAAA,MAAA;EAAA;EAAA;AAAA",
        "names": [
          "value",
          "doStuff",
          "_error",
        ],
        "sourceRoot": undefined,
        "sources": [
          "/path/to/example.ts",
        ],
        "sourcesContent": [
          null,
        ],
        "version": 3,
      },
    },
    "/path/to/example.ts?my-example=1.ts" => {
      "code": "import { server as _entry } from "my-example/server";
import _root from "./example.ts?my-example=0.ts";
export default _entry("26ecc3a5-0-foo", _root);",
      "map": {
        "file": undefined,
        "mappings": "",
        "names": [],
        "sourceRoot": undefined,
        "sources": [],
        "sourcesContent": [],
        "version": 3,
      },
    },
  },
  "map": {
    "file": undefined,
    "mappings": ";AACM,eAAeA,GAAGA,CAAA,EAAG;EACnB,MAAMC,KAAK,GAAG,SAAS;EACvB;IAAA,OAAAC,KAAA,EAAAC,OAAA,UAAAC,OAAA,gDAAAC,OAAA,QADMJ,KAAK;EAAA,EAIT,OAAOK,GAAG,EAAE;IACZC,MAAM,CAACD,GAAG,CAAC;EACb;AACF",
    "names": [
      "foo",
      "value",
      "_type",
      "_result",
      "_$$func",
      "default",
      "err",
      "report",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      async function foo() {
        const value = 'foo bar';
        try {
          'use server';
          await doStuff(value);
        } catch (err) {
          report(err);
        }
      }
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?my-example=0.ts",
  ],
}
`;

exports[`TryStatement > server > should transform valid server try statements 1`] = `
{
  "code": "import { $$func as _$$func } from "dismantle/runtime";
try {
  const [_type, _result] = await _$$func((await import("./example.ts?my-example=1.ts")).default, null)();
} catch (err) {
  report(err);
}",
  "entries": [
    "/path/to/example.ts?my-example=1.ts",
  ],
  "files": Map {
    "/path/to/example.ts?my-example=0.ts" => {
      "code": "export default (async function () {
  try {
    await doStuff();
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
      "map": {
        "file": undefined,
        "mappings": ";;IAGQ,MAAMA,OAAO,CAAC,CAAC;EAAC,SAAAC,MAAA;IAAA,WAAAA,MAAA;EAAA;EAAA;AAAA",
        "names": [
          "doStuff",
          "_error",
        ],
        "sourceRoot": undefined,
        "sources": [
          "/path/to/example.ts",
        ],
        "sourcesContent": [
          null,
        ],
        "version": 3,
      },
    },
    "/path/to/example.ts?my-example=1.ts" => {
      "code": "import { server as _entry } from "my-example/server";
import _root from "./example.ts?my-example=0.ts";
export default _entry("26ecc3a5-0-anonymous", _root);",
      "map": {
        "file": undefined,
        "mappings": "",
        "names": [],
        "sourceRoot": undefined,
        "sources": [],
        "sourcesContent": [],
        "version": 3,
      },
    },
  },
  "map": {
    "file": undefined,
    "mappings": ";AACM;EAAA,OAAAA,KAAA,EAAAC,OAAA,UAAAC,OAAA,gDAAAC,OAAA;AAAA,EAGE,OAAOC,GAAG,EAAE;EACZC,MAAM,CAACD,GAAG,CAAC;AACb",
    "names": [
      "_type",
      "_result",
      "_$$func",
      "default",
      "err",
      "report",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      try {
        'use server';
        await doStuff();
      } catch (err) {
        report(err);
      }
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?my-example=0.ts",
  ],
}
`;
